File: latest_ran_output/claude_sonnet/third_002_naive.js
```javascript
const express = require('express');
const crypto = require('crypto');
const router = express.Router();

// --- Mock Database ---
const orders = new Map([
  ['order_001', { id: 'order_001', status: 'pending', amount: 9999, currency: 'usd', customerId: 'cust_123' }],
  ['order_002', { id: 'order_002', status: 'pending', amount: 4999, currency: 'usd', customerId: 'cust_456' }],
]);

// --- Configuration ---
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your_webhook_secret_key_here';
const ALLOWED_EVENT_TYPES = [
  'payment.succeeded',
  'payment.failed',
  'payment.refunded',
  'payment.disputed',
  'payment.pending',
  'charge.succeeded',
  'charge.failed',
  'charge.refunded',
];

// --- Webhook Signature Verification ---
const verifyWebhookSignature = (rawBody, signature, secret) => {
  if (!signature || !rawBody) {
    return { valid: false, error: 'Missing signature or body' };
  }

  try {
    // Support both simple HMAC and timestamp-based verification (e.g., Stripe style)
    const parts = signature.split(',');
    let timestamp = null;
    let receivedSignature = null;

    if (parts.length > 1) {
      // Timestamp-based: "t=1234567890,v1=abc123..."
      for (const part of parts) {
        const [key, value] = part.split('=');
        if (key === 't') timestamp = value;
        if (key === 'v1') receivedSignature = value;
      }

      if (!timestamp || !receivedSignature) {
        return { valid: false, error: 'Invalid signature format' };
      }

      // Reject webhooks older than 5 minutes to prevent replay attacks
      const currentTime = Math.floor(Date.now() / 1000);
      if (Math.abs(currentTime - parseInt(timestamp)) > 300) {
        return { valid: false, error: 'Webhook timestamp is too old' };
      }

      const signedPayload = `${timestamp}.${rawBody}`;
      const expectedSignature = crypto
        .createHmac('sha256', secret)
        .update(signedPayload, 'utf8')
        .digest('hex');

      const isValid = crypto.timingSafeEqual(
        Buffer.from(expectedSignature, 'hex'),
        Buffer.from(receivedSignature, 'hex')
      );

      return { valid: isValid, error: isValid ? null : 'Signature mismatch' };
    } else {
      // Simple HMAC verification
      const expectedSignature = crypto
        .createHmac('sha256', secret)
        .update(rawBody, 'utf8')
        .digest('hex');

      const receivedSig = signature.startsWith('sha256=')
        ? signature.slice(7)
        : signature;

      const isValid = crypto.timingSafeEqual(
        Buffer.from(expectedSignature, 'hex'),
        Buffer.from(receivedSig, 'hex')
      );

      return { valid: isValid, error: isValid ? null : 'Signature mismatch' };
    }
  } catch (err) {
    return { valid: false, error: `Signature verification error: ${err.message}` };
  }
};

// --- Order Status Mapping ---
const mapEventToOrderStatus = (eventType) => {
  const statusMap = {
    'payment.succeeded': 'paid',
    'payment.failed': 'payment_failed',
    'payment.refunded': 'refunded',
    'payment.disputed': 'disputed',
    'payment.pending': 'payment_pending',
    'charge.succeeded': 'paid',
    'charge.failed': 'payment_failed',
    'charge.refunded': 'refunded',
  };
  return statusMap[eventType] || null;
};

// --- Order Update Logic ---
const updateOrderStatus = async (orderId, newStatus, eventData) => {
  if (!orders.has(orderId)) {
    return { success: false, error: `Order ${orderId} not found` };
  }

  const order = orders.get(orderId);
  const previousStatus = order.status;

  // Define valid status transitions
  const validTransitions = {
    pending: ['paid', 'payment_failed', 'payment_pending'],
    payment_pending: ['paid', 'payment_failed'],
    paid: ['refunded', 'disputed'],
    payment_failed: ['paid', 'payment_pending'],
    disputed: ['paid', 'refunded'],
    refunded: [],
  };

  const allowedTransitions = validTransitions[previousStatus] || [];
  if (!allowedTransitions.includes(newStatus)) {
    return {
      success: false,
      error: `Invalid status transition from '${previousStatus}' to '${newStatus}'`,
    };
  }

  // Update the order
  const updatedOrder = {
    ...order,
    status: newStatus,
    updatedAt: new Date().toISOString(),
    lastEventType: eventData.type,
    lastEventId: eventData.id,
    paymentDetails: {
      transactionId: eventData.data?.transactionId || eventData.data?.chargeId || null,
      amount: eventData.data?.amount || order.amount,
      currency: eventData.data?.currency || order.currency,
    },
  };

  orders.set(orderId, updatedOrder);

  console.log(`[Webhook] Order ${orderId} status updated: ${previousStatus} → ${newStatus}`);

  return {
    success: true,
    previousStatus,
    newStatus,
    order: updatedOrder,
  };
};

// --- Webhook Event Processor ---
const processWebhookEvent = async (event) => {
  const { type, data } = event;

  if (!ALLOWED_EVENT_TYPES.includes(type)) {
    return { processed: false, reason: `Unhandled event type: ${type}` };
  }

  const orderId = data?.orderId || data?.metadata?.orderId;
  if (!orderId) {
    return { processed: false, reason: 'No orderId found in event data' };
  }

  const newStatus = mapEventToOrderStatus(type);
  if (!newStatus) {
    return { processed: false, reason: `No status mapping for event type: ${type}` };
  }

  const result = await updateOrderStatus(orderId, newStatus, event);

  if (!result.success) {
    return { processed: false, reason: result.error };
  }

  return {
    processed: true,
    orderId,
    previousStatus: result.previousStatus,
    newStatus: result.newStatus,
    order: result.order,
  };
};

// --- Idempotency Check ---
const processedEvents = new Set();

const isEventAlreadyProcessed = (eventId) => {
  if (!eventId) return false;
  return processedEvents.has(eventId);
};

const markEventAsProcessed = (eventId) => {
  if (eventId) {
    processedEvents.add(eventId);
    // Clean up old events after 24 hours (in production, use Redis or DB)
    setTimeout(() => processedEvents.delete(eventId), 24 * 60 * 60 * 1000);
  }
};

// --- Webhook Route ---
// IMPORTANT: Raw body parser must be used before express.json() for this route
// In your main app.js, mount this router BEFORE applying express.json() globally,
// or use express.raw() specifically for this path:
//   app.use('/webhooks', express.raw({ type: 'application/json' }), webhookRouter);

router.post(
  '/payment',
  express.raw({ type: 'application/json' }), // Capture raw body for signature verification
  async (req, res) => {
    const startTime = Date.now();
    const requestId =

File: latest_ran_output/claude_sonnet/third_002_naive.py
```python
import os
import hmac
import hashlib
import json
import logging
from datetime import datetime
from functools import wraps
from flask import Flask, request, jsonify, abort
import jwt
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

app = Flask(__name__)

# Configuration — load from environment variables in production
WEBHOOK_SECRET = os.environ.get("WEBHOOK_SECRET", "your-webhook-secret-key")
API_SECRET_KEY = os.environ.get("API_SECRET_KEY", "your-api-secret-key")
SIGNATURE_HEADER = "X-Payment-Signature"
TIMESTAMP_HEADER = "X-Payment-Timestamp"
TOLERANCE_SECONDS = 300  # 5-minute replay-attack window

# ---------------------------------------------------------------------------
# Simulated in-memory order store (replace with your real DB layer)
# ---------------------------------------------------------------------------
ORDERS: dict[str, dict] = {
    "ORD-001": {"id": "ORD-001", "status": "pending", "amount": 9999, "currency": "USD"},
    "ORD-002": {"id": "ORD-002", "status": "pending", "amount": 4999, "currency": "USD"},
}

# ---------------------------------------------------------------------------
# Valid order status transitions
# ---------------------------------------------------------------------------
VALID_TRANSITIONS: dict[str, list[str]] = {
    "pending":    ["paid", "failed", "cancelled"],
    "paid":       ["refunded", "partially_refunded", "disputed"],
    "failed":     ["pending"],           # allow retry
    "cancelled":  [],
    "refunded":   [],
    "partially_refunded": ["refunded"],
    "disputed":   ["paid", "refunded"],
}

# ---------------------------------------------------------------------------
# Mapping from payment-provider event types to internal order statuses
# ---------------------------------------------------------------------------
EVENT_STATUS_MAP: dict[str, str] = {
    "payment.succeeded":          "paid",
    "payment.failed":             "failed",
    "payment.cancelled":          "cancelled",
    "payment.refunded":           "refunded",
    "payment.partially_refunded": "partially_refunded",
    "payment.dispute.created":    "disputed",
    "payment.dispute.resolved":   "paid",
}

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def verify_webhook_signature(payload: bytes, signature: str, timestamp: str) -> bool:
    """
    Verify HMAC-SHA256 signature sent by the payment provider.

    Expected signature format:  sha256=<hex-digest>
    Signed string:              <timestamp>.<raw-body>
    """
    if not signature or not timestamp:
        logger.warning("Missing signature or timestamp header.")
        return False

    # Reject stale requests (replay-attack protection)
    try:
        request_time = int(timestamp)
        current_time = int(datetime.utcnow().timestamp())
        if abs(current_time - request_time) > TOLERANCE_SECONDS:
            logger.warning("Webhook timestamp outside tolerance window.")
            return False
    except ValueError:
        logger.warning("Invalid timestamp header value.")
        return False

    # Reconstruct the expected signature
    signed_payload = f"{timestamp}.".encode() + payload
    expected_sig = hmac.new(
        WEBHOOK_SECRET.encode(),
        signed_payload,
        hashlib.sha256
    ).hexdigest()

    # Strip optional "sha256=" prefix supplied by some providers
    received_sig = signature.removeprefix("sha256=")

    return hmac.compare_digest(expected_sig, received_sig)


def require_webhook_auth(f):
    """Decorator that enforces webhook signature verification."""
    @wraps(f)
    def decorated(*args, **kwargs):
        raw_body = request.get_data()  # must read before parsing JSON
        signature = request.headers.get(SIGNATURE_HEADER, "")
        timestamp = request.headers.get(TIMESTAMP_HEADER, "")

        if not verify_webhook_signature(raw_body, signature, timestamp):
            logger.error("Webhook signature verification failed.")
            abort(401, description="Invalid or missing webhook signature.")

        return f(*args, **kwargs)
    return decorated


def get_order(order_id: str) -> dict | None:
    """Fetch an order by ID (replace with real DB call)."""
    return ORDERS.get(order_id)


def update_order_status(order_id: str, new_status: str) -> dict:
    """
    Persist the new order status (replace with real DB write).
    Returns the updated order record.
    """
    ORDERS[order_id]["status"] = new_status
    ORDERS[order_id]["updated_at"] = datetime.utcnow().isoformat()
    logger.info("Order %s updated to status '%s'.", order_id, new_status)
    return ORDERS[order_id]


def is_valid_transition(current_status: str, new_status: str) -> bool:
    """Check whether a status transition is permitted."""
    return new_status in VALID_TRANSITIONS.get(current_status, [])


# ---------------------------------------------------------------------------
# Webhook endpoint
# ---------------------------------------------------------------------------

@app.route("/webhooks/payment", methods=["POST"])
@require_webhook_auth
def payment_webhook():
    """
    Receive and process payment-provider webhook events.

    Expected JSON body:
    {
        "event_type": "payment.succeeded",
        "event_id":   "evt_abc123",
        "created_at": 1700000000,
        "data": {
            "order_id":        "ORD-001",
            "payment_id":      "pay_xyz789",
            "amount":          9999,
            "currency":        "USD",
            "failure_reason":  null   // optional
        }
    }
    """
    # Parse JSON safely
    try:
        payload = request.get_json(force=True, silent=False)
        if payload is None:
            raise ValueError("Empty or non-JSON body.")
    except Exception as exc:
        logger.error("Failed to parse webhook payload: %s", exc)
        return jsonify({"error": "Invalid JSON payload."}), 400

    # Validate required top-level fields
    event_type = payload.get("event_type")
    event_id   = payload.get("event_id")
    data       = payload.get("data", {})

    if not event_type or not event_id or not isinstance(data, dict):
        logger.warning("Malformed webhook payload: %s", payload)
        return jsonify({"error": "Missing required fields: event_type, event_id, data."}), 400

    # Validate required data fields
    order_id   = data.get("order_id")
    payment_id = data.get("payment_id")

    if not order_id or not payment_id:
        logger.warning("Webhook data missing order_id or payment_id.")
        return jsonify({"error": "Missing required data fields: order_id, payment_id."}), 400

    logger.info(
        "Received event '%s' (id=%s) for order '%s', payment '%s'.",
        event_type, event_id, order_id, payment_id
    )

    # Resolve target status from event type
    new_status = EVENT_STATUS_MAP.get(event_type)
    if new_status is None:
        logger.info("Unhandled event type '%s' — acknowledged without action.", event_type)
        return jsonify({"received": True, "message": f"Event type '{event_type}' not handled."}), 200

    # Retrieve the order
    order = get_order(order_id)
    if order is None:
        logger.error("Order '%s' not found.", order_id)
        return jsonify({"error": f"Order '{order_id}' not found."}), 404

    current_status = order["status"]

    # Idempotency guard — already in target status
    if current_status == new_status:
        logger